File: src/main/java/com/github/lppedd/inspections/unnecessaryCharSequenceToString/UnnecessaryCharSequenceToStringElementVisitor.java
package com.github.lppedd.inspections.unnecessaryCharSequenceToString;

import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.github.lppedd.inspections.InspectionsBundle;
import com.intellij.codeInsight.daemon.impl.analysis.HighlightControlFlowUtil;
import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.openapi.progress.ProgressIndicatorProvider;
import com.intellij.psi.*;
import com.intellij.psi.search.searches.ReferencesSearch;
import com.intellij.psi.util.PsiUtil;
import com.siyeh.ig.callMatcher.CallMatcher;
import com.siyeh.ig.psiutils.ExpressionUtils;
import com.siyeh.ig.psiutils.MethodCallUtils;
import com.siyeh.ig.psiutils.TypeUtils;

/**
 * @author Edoardo Luppi
 */
class UnnecessaryCharSequenceToStringElementVisitor extends JavaElementVisitor {
  private static final String CHAR_SEQUENCE_FQN = "java.lang.CharSequence";
  private static final String STRING_FQN = "java.lang.String";
  private static final CallMatcher TO_STRING =
      CallMatcher.exactInstanceCall(CHAR_SEQUENCE_FQN, "toString");

  private final @NotNull ProblemsHolder problemsHolder;
  private final boolean isOnTheFly;

  UnnecessaryCharSequenceToStringElementVisitor(
      final @NotNull ProblemsHolder problemsHolder,
      final boolean isOnTheFly) {
    this.problemsHolder = problemsHolder;
    this.isOnTheFly = isOnTheFly;
  }

  @Override
  public void visitField(final @NotNull PsiField psiField) {
    if (psiField.hasModifierProperty(PsiModifier.PRIVATE)) {
      ProgressIndicatorProvider.checkCanceled();
      addDescriptors(checkVariable(psiField));
    }
  }

  @Override
  public void visitLocalVariable(final @NotNull PsiLocalVariable psiLocalVariable) {
    ProgressIndicatorProvider.checkCanceled();
    addDescriptors(checkVariable(psiLocalVariable));
  }

  private void addDescriptors(final @NotNull ProblemDescriptor[] descriptors) {
    for (final var descriptor : descriptors) {
      problemsHolder.registerProblem(descriptor);
    }
  }

  @NotNull ProblemDescriptor[] checkVariable(final @NotNull PsiVariable psiVariable) {
    // We consider only variables which are final or never reassigned,
    // and of type CharSequence/String (because we are looking for toString() calls
    // and a String can be assigned only to a String or CharSequence)
    if (isVariableTypeStringOrCharSequence(psiVariable.getType()) &&
        isVariableEffectivelyFinal(psiVariable) &&
        shouldSearchForToStringCalls(psiVariable)) {
      final var initializer = psiVariable.getInitializer();
      return checkVariableRecursive(psiVariable, initializer).toArray(ProblemDescriptor[]::new);
    }

    return ProblemDescriptor.EMPTY_ARRAY;
  }

  private @NotNull Stream<ProblemDescriptor> checkVariableRecursive(
      final @NotNull PsiVariable variable,
      final @Nullable PsiExpression initializer) {
    // noinspection SimplifiableIfStatement
    if (initializer == null) {
      return Stream.empty();
    }

    return ExpressionUtils
        .nonStructuralChildren(initializer)
        .filter(PsiMethodCallExpression.class::isInstance)
        .map(PsiMethodCallExpression.class::cast)
        .flatMap(psiMethodCallExpression -> {
          final var psiMethodExpression = psiMethodCallExpression.getMethodExpression();
          final var psiQualifierExpression = psiMethodExpression.getQualifierExpression();
          final var childProblems = checkVariableRecursive(variable, psiQualifierExpression);
          return TO_STRING.test(psiMethodCallExpression)
              ? Stream.concat(Stream.of(buildProblem(variable, psiMethodExpression)), childProblems)
              : childProblems;
        });
  }

  private @NotNull ProblemDescriptor buildProblem(
      final @NotNull PsiVariable variable,
      final @NotNull PsiReferenceExpression methodExpression) {
    return problemsHolder.getManager().createProblemDescriptor(
        Objects.requireNonNull(methodExpression.getReferenceNameElement()),
        InspectionsBundle.get("unnecessary.charsequence.tostring.displayName"),
        true,
        ProblemHighlightType.LIKE_UNUSED_SYMBOL,
        isOnTheFly,
        new UnnecessaryCharSequenceToStringFix(this, variable)
    );
  }

  private static boolean isVariableTypeStringOrCharSequence(final PsiType psiVariableType) {
    return TypeUtils.typeEquals(STRING_FQN, psiVariableType) ||
           TypeUtils.typeEquals(CHAR_SEQUENCE_FQN, psiVariableType);
  }

  private static boolean isVariableEffectivelyFinal(final @NotNull PsiVariable psiVariable) {
    final var hasFinalModifier = psiVariable.hasModifierProperty(PsiModifier.FINAL);

    if (hasFinalModifier) {
      return true;
    }

    if (psiVariable instanceof PsiField) {
      return false;
    }

    final var scope = PsiUtil.getVariableCodeBlock(psiVariable, null);
    return scope != null && HighlightControlFlowUtil.isEffectivelyFinal(psiVariable, scope, null);
  }

  private static boolean shouldSearchForToStringCalls(final @NotNull PsiVariable psiVariable) {
    // We search for all the usages of the variable.
    // If every method parameter where the variable is used as argument accepts
    // a CharSequence, then we can delete the toString() call
    final var psiReferences =
        ReferencesSearch.search(psiVariable, psiVariable.getUseScope())
                        .findAll()
                        .stream()
                        .filter(PsiExpression.class::isInstance)
                        .map(PsiExpression.class::cast)
                        .collect(Collectors.toList());

    if (psiReferences.isEmpty()) {
      // No usages found so we can safely try to simplify code
      return true;
    }

    for (final var psiVariableReference : psiReferences) {
      var psiVariableReferenceContext = psiVariableReference.getContext();

      if (psiVariableReferenceContext instanceof PsiReferenceExpression psiReferenceExpression &&
          psiReferenceExpression.getContext() instanceof PsiMethodCallExpression psiMethodCallExpr) {
        // The variable is used to call a method, e.g. myVar.someMethod()
        // We need to check if the method is inherited from CharSequence,
        // and only in this case allow removing toString()
        final var psiMethod = psiMethodCallExpr.resolveMethod();

        if (psiMethod == null) {
          continue;
        }

        final var psiMethodContainingClass = Objects.requireNonNull(psiMethod.getContainingClass());

        if (CHAR_SEQUENCE_FQN.equals(psiMethodContainingClass.getQualifiedName())) {
          // The variable is used as CharSequence
          psiVariableReferenceContext = psiMethodCallExpr.getParent();
        } else {
          final var superMethods = psiMethod.findDeepestSuperMethods();

          if (superMethods.length == 0) {
            if (!CHAR_SEQUENCE_FQN.equals(psiMethodContainingClass.getQualifiedName())) {
              return false;
            }
          } else {
            final var containingClass = Objects.requireNonNull(superMethods[0].getContainingClass());

            if (!CHAR_SEQUENCE_FQN.equals(containingClass.getQualifiedName())) {
              return false;
            }

            psiVariableReferenceContext = psiMethodCallExpr.getParent();
          }
        }
      }

      if (psiVariableReferenceContext instanceof PsiExpressionList psiExpressionList &&
          psiExpressionList.getContext() instanceof PsiMethodCallExpression psiMethodCallExpr &&
          !isMethodParameterCharSequence(psiVariableReference, psiMethodCallExpr)) {
        return false;
      }
    }

    return true;
  }

  private static boolean isMethodParameterCharSequence(
      final @NotNull PsiExpression psiArgumentExpression,
      final @NotNull PsiMethodCallExpression psiMethodCallExpression) {
    final var parameterForArgument = MethodCallUtils.getParameterForArgument(psiArgumentExpression);

    if (parameterForArgument != null) {
      final var parameterType = parameterForArgument.getType().getDeepComponentType();
      return TypeUtils.typeEquals(CHAR_SEQUENCE_FQN, parameterType);
    }

    // Might be a vararg argument
    final var psiVarargMethod = psiMethodCallExpression.resolveMethod();

    if (psiVarargMethod == null) {
      return false;
    }

    final var psiParameters = psiVarargMethod.getParameterList().getParameters();

    if (psiParameters.length == 0) {
      return false;
    }

    final var psiParameter = psiParameters[psiParameters.length - 1];
    return psiParameter.isVarArgs() &&
           TypeUtils.typeEquals(CHAR_SEQUENCE_FQN, psiParameter.getType().getDeepComponentType());
  }
}

File: src/main/kotlin/com/github/lppedd/inspections/unnecessaryCharSequenceToString/UnnecessaryCharSequenceToStringElementVisitor.kt
package com.github.lppedd.inspections.unnecessaryCharSequenceToString

import com.github.lppedd.inspections.*
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.codeInspection.ProblemHighlightType.LIKE_UNUSED_SYMBOL
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.openapi.progress.ProgressIndicatorProvider
import com.intellij.psi.*
import com.intellij.psi.search.searches.ReferencesSearch
import com.siyeh.ig.callMatcher.CallMatcher
import com.siyeh.ig.psiutils.ExpressionUtils
import java.util.stream.Stream
import kotlin.streams.toList

private const val STRING_FQN = "java.lang.String"
private const val CHAR_SEQUENCE_FQN = "java.lang.CharSequence"
private val TO_STRING_MATCHER: CallMatcher = CallMatcher.exactInstanceCall(CHAR_SEQUENCE_FQN, "toString")

/**
 * @author Edoardo Luppi
 */
internal class UnnecessaryCharSequenceToStringElementVisitor(
  private val problemsHolder: ProblemsHolder,
  private val isOnTheFly: Boolean
) : JavaElementVisitor() {
  override fun visitField(psiField: PsiField) {
    if (psiField.isPrivateAndFinal()) {
      ProgressIndicatorProvider.checkCanceled()
      addDescriptors(checkVariable(psiField))
    }
  }

  override fun visitLocalVariable(psiLocalVariable: PsiLocalVariable) {
    ProgressIndicatorProvider.checkCanceled()
    addDescriptors(checkVariable(psiLocalVariable))
  }

  private fun addDescriptors(descriptors: List<ProblemDescriptor>) {
    descriptors.forEach(problemsHolder::registerProblem)
  }

  fun checkVariable(psiVariable: PsiVariable): List<ProblemDescriptor> =
    if (psiVariable.type.isOneOf(STRING_FQN, CHAR_SEQUENCE_FQN) &&
        psiVariable.isEffectivelyFinal() &&
        psiVariable.isUsedAsCharSequenceOnly()) {
      checkVariableRecursive(psiVariable, psiVariable.initializer).toList()
    } else {
      emptyList()
    }

  private fun PsiVariable.isUsedAsCharSequenceOnly(): Boolean {
    // We search for all the usages of the variable.
    // If every method parameter where the variable is used as argument accepts
    // a CharSequence, then we can delete the toString() call
    val psiReferences =
      ReferencesSearch.search(this, useScope)
        .findAll()
        .filterIsInstance(PsiExpression::class.java)
        .toList()
        .ifEmpty {
          // No usages found so we can safely simplify code
          return true
        }

    for (psiVariableReference in psiReferences) {
      var varReferenceContext = psiVariableReference.context

      // The reference is used to call a method.
      // We need to check if that called method is owned by CharSequence or not
      if (varReferenceContext is PsiReferenceExpression) {
        val methodCallExpression = varReferenceContext.context as? PsiMethodCallExpression
        val method = methodCallExpression?.resolveMethod() ?: return false

        varReferenceContext =
          if (method.containingClass?.equalsToText(CHAR_SEQUENCE_FQN) == true) {
            // The called method is owned by CharSequence
            methodCallExpression
          } else {
            // The called method is not owned by CharSequence,
            // but it could have been overridden so we need to "walk" down to super
            if (!method.getDeepestSuperMethodsClasses().any { CHAR_SEQUENCE_FQN == it?.qualifiedName }) {
              return false
            }

            methodCallExpression.parent
          }
      }

      // The reference is used inside a method call, someVar.someMethod(|varRef|, ...)
      // We need to check if the corresponding method's parameter is of type CharSequence
      if (varReferenceContext is PsiExpressionList &&
          (varReferenceContext.context as? PsiMethodCallExpression)
            ?.getParameter(psiVariableReference)
            ?.deepComponentType
            ?.equalsToText(CHAR_SEQUENCE_FQN) == false) {
        return false
      }
    }

    return true
  }

  private fun checkVariableRecursive(
    variable: PsiVariable,
    initializer: PsiExpression?
  ): Stream<ProblemDescriptor> {
    return if (initializer == null) Stream.empty()
    else ExpressionUtils
      .nonStructuralChildren(initializer)
      .filterIsInstance(PsiMethodCallExpression::class.java)
      .flatMap { psiMethodCallExpression ->
        val psiMethodExpression = psiMethodCallExpression.methodExpression
        val psiQualifierExpression = psiMethodExpression.qualifierExpression
        val childProblems = checkVariableRecursive(variable, psiQualifierExpression)

        if (TO_STRING_MATCHER.test(psiMethodCallExpression)) {
          Stream.of(buildProblem(variable, psiMethodExpression)) + childProblems
        } else {
          childProblems
        }
      }
  }

  private fun buildProblem(
    variable: PsiVariable,
    methodExpression: PsiReferenceExpression
  ): ProblemDescriptor =
    problemsHolder.manager.createProblemDescriptor(
      methodExpression.referenceNameElement!!,
      InspectionsBundle["unnecessary.charsequence.tostring.displayName"],
      true,
      LIKE_UNUSED_SYMBOL,
      isOnTheFly,
      UnnecessaryCharSequenceToStringFix(this, variable)
    )
}

private fun PsiField.isPrivateAndFinal(): Boolean =
  modifierList?.run {
    hasModifierProperty(PsiModifier.PRIVATE) &&
        hasModifierProperty(PsiModifier.FINAL)
  } ?: false

private fun PsiExpression.argumentIndex(): Int {
  return (parent as? PsiExpressionList ?: return -1).expressions.indexOf(this)
}

private fun PsiMethodCallExpression.getParameter(psiExpression: PsiExpression): PsiParameter? {
  val method = resolveMethod() ?: return null
  val parameters = method.parameterList.parameters
  val argumentIndex = psiExpression.argumentIndex()

  return when {
    // Something wrong?
    argumentIndex < 0 -> null
    // Argument passed to vararg method's parameter
    argumentIndex > parameters.lastIndex && method.isVarArgs -> parameters.lastOrNull()
    // Argument passed to normal method's parameter
    else -> parameters[argumentIndex]
  }
}
